### PR TITLE
Iox 1942 introduce semantic string

### LINKED ISSUE #1942: Strong string types
## Brief feature description

In multiple constructs we use a `iox::string` to represent a posix user/group, filename, path, filepath etc. This leads to a lot of code duplication since the validity of the string has to be verified on multiple levels multiple times.
Furthermore, it makes it unclear to the user what kind of string-type is required and what restriction/contract it shall follow.

## Detailed information

* [x] Introduce a `SemanticString` based on a `iox::String` which is used as base for strings with a contract.
* [ ] Implement `User`, `Group`, `FileName`, `FilePath`, `Path` as semantic strings.

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [ ] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code according to our coding style and naming conventions
- [ ] Unit tests have been written for new behavior
- [ ] Public API changes are documented via doxygen
- [ ] Copyright owner are updated in the changed files
- [ ] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [ ] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

- Closes #1942

### CODE REVIEW

File: iceoryx_hoofs/vocabulary/include/iox/semantic_string.hpp [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    /// @return string reference containing the actual value.
+    constexpr const string<Capacity>& as_string() const noexcept;
+
+    /// @brief Appends another string to the SemanticString. If the value contains
+    ///        invalid characters or the result would end up in invalid content
+    ///        it fails.
+    /// @param[in] value the value which should be added
+    /// @return on failure the error inside the expected describes the failure
+    template <typename T>
+    expected<SemanticStringError> append(const T& value) noexcept;
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/vocabulary/include/iox/semantic_string.hpp:118]: I wonder if we should restrict `T` here via `IsStringOrCharArray`.

Comment (elfenpiff) [reply iceoryx_hoofs/vocabulary/include/iox/semantic_string.hpp:118]: I could do this but then the code would look ugly and it is nevertheless handled by the underlying string.
When I solve this at this level the compiler failures are also not more readable - I already tried it.

**Review by elBoberido — COMMENTED** [review]
Looks good. The tests are not yet reviewed.

File: iceoryx_hoofs/vocabulary/include/iox/semantic_string.hpp [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    /// @brief checks if another SemanticString is equal to this string
+    /// @param [in] rhs the other SemanticString
+    /// @return true if the contents are equal, otherwise false
+    bool operator==(const SemanticString& rhs) const noexcept;
+
+    /// @brief checks if another string or char array is equal to this string
+    /// @param [in] rhs the other string
+    /// @return true if the contents are equal, otherwise false
+    template <typename T>
+    IsStringOrCharArray<T, bool> operator==(const T& rhs) const noexcept;
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/vocabulary/include/iox/semantic_string.hpp:139]: It would be something like
```cpp
if (username1 == filename2) {
   // do something with username2
}
```

[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_hoofs/vocabulary/include/iox/semantic_string.hpp:139]: I will think and sleep about it.

[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/vocabulary/include/iox/semantic_string.hpp:139]: I was just a food for thought with the idea that it is always easier to add than to take away. No strong opinion in either direction.

[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_hoofs/vocabulary/include/iox/semantic_string.hpp:139]: The more I think about it the more I like it when the comparison works out of the box. I see all over the place use cases for comparing different semantic strings. And the beauty of the semantic string is that it cannot be easily mixed up with the actual functions which do things ...
Of course, the problem you drafted is a problem but if feels weird since the user mixes up variable names and this can happen also at any time any where and I would not try to improve the safety for this ... this is a task without end and not too much gain.

File: iceoryx_hoofs/test/moduletests/test_vocabulary_semantic_string.cpp [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+}
+
+TYPED_TEST(SemanticString_test, AppendInvalidContentToValidStringFails)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "fddf4a56-c368-4ff0-8727-e732d6ebc87f");
+    using SutType = typename TestFixture::SutType;
+
+    for (auto& value : TestValues<SutType>::VALID_VALUES)
+    {
+        for (auto& invalid_value : TestValues<SutType>::INVALID_CHARACTER_VALUES)
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_vocabulary_semantic_string.cpp:182]: The test name does not match with the test values. `InvalidContent` vs `INVALID_CHARACTER`. I think you need to duplicate this test case to have both cases tested.

File: iceoryx_hoofs/test/moduletests/test_vocabulary_semantic_string.cpp [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+        {
+            auto sut = SutType::create(string<SutType::capacity()>(TruncateToCapacity, value.c_str()));
+            ASSERT_THAT(sut.has_error(), Eq(false));
+
+            EXPECT_THAT(sut->append(string<SutType::capacity()>(TruncateToCapacity, invalid_value.c_str())).has_error(),
+                        Eq(true));
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_vocabulary_semantic_string.cpp:188]: Can you also please check for the correct error code.

File: iceoryx_hoofs/test/moduletests/test_vocabulary_semantic_string.cpp [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+
+                auto result = value;
+                result.insert(insert_position, add_value);
+                EXPECT_THAT(sut->as_string().c_str(), StrEq(result));
+            }
+        }
+    }
+}
+
+TYPED_TEST(SemanticString_test, InsertInvalidContentToValidStringFails)
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_vocabulary_semantic_string.cpp:251]: This is similar to `AppendInvalidContentToValidStringFails` with `InvalidContent` in the test name and `INVALID_CHARACTER` for test values. Since `append` uses `insert` I don't think it is required to duplicate the test but only to adjust the names.

File: iceoryx_hoofs/test/moduletests/test_vocabulary_semantic_string.cpp [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+            for (uint64_t insert_position = 0; insert_position < value.size(); ++insert_position)
+            {
+                auto sut = SutType::create(string<SutType::capacity()>(TruncateToCapacity, value.c_str()));
+                ASSERT_THAT(sut.has_error(), Eq(false));
+
+                EXPECT_THAT(sut->insert(insert_position,
+                                        string<SutType::capacity()>(TruncateToCapacity, add_value.c_str()),
+                                        add_value.size())
+                                .has_error(),
+                            Eq(true));
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_vocabulary_semantic_string.cpp:269]: Please also check for the correct error code.

File: iceoryx_hoofs/test/moduletests/test_vocabulary_semantic_string.cpp [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+
+TYPED_TEST(SemanticString_test, GreaterThanOperatorWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "8c046cff-fb69-43b4-9a45-e86f17c874db");
+
+    EXPECT_FALSE(this->greater_value > this->greater_value);
+    EXPECT_FALSE(this->smaller_value > this->greater_value);
+    EXPECT_TRUE(this->greater_value > this->smaller_value);
+}
+} // namespace
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_vocabulary_semantic_string.cpp:367]: Should methods like `size`, `capacity` and `as_string` also be explicitly tested? I mean not all corner cases, just to check whether the call is correctly forwarded to the underlying `iox::string`.

**Review by elBoberido — COMMENTED** [review]
Just a few minor thing.

File: iceoryx_hoofs/test/moduletests/test_vocabulary_semantic_string.cpp [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+    {"some-!user"}, {"*kasjd"}, {"_fuuuas"}, {"asd/asd"}, {";'1'fuuuu"}, {"argh/"}, {"fuu/arg/bla"}};
+template <>
+std::vector<std::string> TestValues<UserName>::INVALID_CONTENT_VALUES{
+    {""}, {"-do-not-start-with-dash"}, {"5do-not-start-with-a-number"}};
+template <>
+std::vector<std::string> TestValues<UserName>::TOO_LONG_CONTENT_VALUES{{"i-am-waaaaay-toooooooo-loooooooong"}};
+template <>
+std::string TestValues<UserName>::GREATER_VALID_VALUE{"zebra-zusel"};
+template <>
+std::string TestValues<UserName>::SMALLER_VALID_VALUE{"alfons-alf"};
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_hoofs/test/moduletests/test_vocabulary_semantic_string.cpp:60]: And he was a dentist

File: iceoryx_hoofs/test/moduletests/test_vocabulary_semantic_string.cpp (hunk 3 above)
Comment (elfenpiff) [reply iceoryx_hoofs/test/moduletests/test_vocabulary_semantic_string.cpp:182]: Here and also in the insert test case I adjusted the test name to `AppendInvalidCharactersToValidStringFails`

The reason is that adding invalid characters will fail but adding invalid content most likely will succeed. Here an example:
1. A file name is not allowed to contain `.` or `..` but adding `.` to `hello` results in the valid file name `hello.`
2. Users are not allowed to start with dash, so `-shumway` is illegal content but adding or inserting this content results in the legal name `gordon-shumway`.

Comment (elBoberido) [reply iceoryx_hoofs/test/moduletests/test_vocabulary_semantic_string.cpp]: Ah, okay. It is just that the path is not tested according to codecov

File: iceoryx_examples/iceoptions/iox_publisher_with_options.cpp [hunk 9]
```diff
@@ -48,7 +48,7 @@ int main()
 
     //  we allow the subscribers to block the publisher if they want to ensure that no samples are lost
     //! [too slow policy]
-    publisherOptions.subscriberTooSlowPolicy = iox::popo::ConsumerTooSlowPolicy::WAIT_FOR_CONSUMER;
+    publisherOptions.subscriberTooSlowPolicy = iox::popo::ConsumerTooSlowPolicy::DISCARD_OLDEST_DATA;
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_examples/iceoptions/iox_publisher_with_options.cpp]: It never happened!

**Review by elBoberido — APPROVED** [review]
